Content.Shared/_Crescent/Hardpoints/GridHardpointLimiterSystem.cs

```diff
@@ -0,0 +1,221 @@
+using Content.Shared._Crescent.Hardpoints;
+using Content.Shared.Construction.Components;
+using Robust.Shared.GameObjects;
+using Robust.Shared.Map;
+using Robust.Shared.Log;
+using Robust.Shared.IoC;
+using Robust.Shared.Timing;
+using Robust.Shared.Map.Components;
+
+namespace Content.Server._Crescent.Hardpoints;
+
+public sealed class GridHardpointLimiterSystem : EntitySystem
+{
+    [Dependency] private readonly IMapManager _mapManager = default!;
+    [Dependency] private readonly SharedTransformSystem _xform = default!;
+    [Dependency] private readonly IGameTiming _timing = default!; // Fixed: Use SharedTransformSystem
+
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<HardpointCannonAnchoredEvent>(OnCannonAnchored);
+        SubscribeLocalEvent<HardpointCannonDeanchoredEvent>(OnCannonDeanchored);
+
+        // Instead of GridInitializeEvent, let's use MapInitEvent on grids
+        SubscribeLocalEvent<MapGridComponent, MapInitEvent>(OnGridMapInit);
+
+        // IMPORTANT: Listen for hardpoint anchoring attempts
+        SubscribeLocalEvent<HardpointComponent, AnchorAttemptEvent>(OnHardpointAnchorAttempt);
+        SubscribeLocalEvent<HardpointComponent, UnanchorAttemptEvent>(OnHardpointUnanchorAttempt);
+
+        // CRITICAL: Final check right before anchoring completes
+        SubscribeLocalEvent<HardpointComponent, BeforeAnchoredEvent>(OnHardpointBeforeAnchored);
+    }
+
+    private void OnGridMapInit(EntityUid uid, MapGridComponent component, MapInitEvent args)
+    {
+        // Add a small delay to ensure all entities are loaded
+        Timer.Spawn(200, () => InitializeGridLimits(uid));
+    }
```

Comment:
No, figure out the right event for this , likely will be OnGridInitialized() or LateGridInitialize() , or just do it like the mass adjuster by speed component does.